File: capacity-club-api/src/domain-modules/exercise-training/model/payload/exercise-training-create.payload.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsEnum, IsNotEmpty, IsOptional, Length } from 'class-validator';
import { TrainingCircuit } from 'domain-modules/training-circuit/model/entity/training-circuit.entity';
import { TrainingIntensity } from '../enum';
import { ExerciseData } from 'domain-modules/exercise-data/model';

/**
 * Class used to create exercise training.
 * This class defines the payload used when creating new exercise training.
 */
export class ExerciseTrainingCreatePayload {
  @ApiProperty()
  @Length(1, 50)
  @IsNotEmpty()
  nb_reps: number;

  @ApiProperty()
  @IsNotEmpty()
  intensity: string;

  @ApiProperty()
  @IsEnum(TrainingIntensity)
  @IsNotEmpty()
  intensityType: TrainingIntensity;

  @ApiProperty({ type: () => ExerciseData })
  exercise_data: ExerciseData;

  @ApiProperty({ type: () => TrainingCircuit })
  training_circuit: TrainingCircuit;
}

File: capacity_club_mobile_app/lib/domain-features/data/exercise-training/model/payload/exercise_training_create_payload.dart
import 'package:capacity_club_mobile_app/core/model/abstract/payload.dart';
import 'package:capacity_club_mobile_app/domain-features/data/exercise-training/model/enum/training_intensity.dart';
import 'package:capacity_club_mobile_app/domain-features/data/exercise-data/model/exercise_data_model.dart';
import 'package:capacity_club_mobile_app/domain-features/data/training-circuit/model/training_circuit_model.dart';
import 'package:equatable/equatable.dart';

class ExerciseTrainingCreatePayload extends Equatable implements Payload {
  final int nb_reps;
  final String intensity;
  final TrainingIntensity intensityType;
  final ExerciseDataModel exercise_data;
  final TrainingCircuitModel training_circuit;

  const ExerciseTrainingCreatePayload({
    required this.nb_reps,
    required this.intensity,
    required this.intensityType,
    required this.exercise_data,
    required this.training_circuit,
  });

  Map<String, dynamic> toJson() => {
        'nb_reps': nb_reps,
        'intensity': intensity,
        'intensityType': intensityType.toString(), // Convert enum to String
        'exercise_data': exercise_data.toJson(), // Use model's toJson method
        'training_circuit':
            training_circuit.toJson(), // Use model's toJson method
      };

  @override
  List<Object?> get props => [
        nb_reps,
        intensity,
        intensityType,
        exercise_data,
        training_circuit,
      ];
}
